### PR TITLE
Fix hamiltonian casting bug

### PR DESCRIPTION
**Pull request by puzzleshark**
**Context**

A Hamiltonian object holds two lists: a list of coefficients and a list of terms. A Hamiltonian object may be multiplied by a scalar value.

There's an issue, however, when a floating scalar is used to multiply a Hamiltonian that has integer coefficients because the scalar itself is being cast to an integer. The behaviour is more general and is due to the use of `qml.math.cast_like` that casts the dtype of the target.

**Changes**

Removes the `cast_like` call from the `__mul__` and `__imul__` methods of `Hamiltonian`.

**Related issue**
Closes https://github.com/PennyLaneAI/pennylane/issues/2361.

### CODE REVIEW

File: pennylane/ops/qubit/hamiltonian.py [hunk 1]
```diff
@@ -567,7 +567,7 @@ def __mul__(self, a):
         r"""The scalar multiplication operation between a scalar and a Hamiltonian."""
         if isinstance(a, (int, float)):
             self_coeffs = copy(self.coeffs)
-            coeffs = qml.math.multiply(qml.math.cast_like([a], self_coeffs), self_coeffs)
+            coeffs = a * self_coeffs
```
Comment (antalszava) [inline pennylane/ops/qubit/hamiltonian.py:570]: Hi @puzzleshark, how does this solution work with interfaces (TF, Torch, JAX)?

Comment (puzzleshark) [reply pennylane/ops/qubit/hamiltonian.py:570]: there is a check that a must be a int or float, so I don't think there can be a problem

Comment (puzzleshark) [reply pennylane/ops/qubit/hamiltonian.py]: changed to `qml.math.multiply` @antalszava

File: pennylane/ops/qubit/hamiltonian.py [hunk 2]
```diff
@@ -601,7 +601,7 @@ def __iadd__(self, H):
     def __imul__(self, a):
         r"""The inplace scalar multiplication operation between a scalar and a Hamiltonian."""
         if isinstance(a, (int, float)):
-            self._coeffs = qml.math.multiply(qml.math.cast_like([a], self._coeffs), self._coeffs)
+            self._coeffs = qml.math.multiply(a, self._coeffs)
```
Comment (antalszava) [inline pennylane/ops/qubit/hamiltonian.py:604]: The same issue seems to be relevant for the in-place multiplication operator too.

**Review by antalszava — APPROVED** [review]
:100:

**Pull request merged**